### PR TITLE
Add dynamic log level change support

### LINKED ISSUE #6213: Add native support for modifying the log level programmatically at runtime
## Summary

We need a new Java API to change the log level programmatically at runtime without restarting the application. At the moment, the only way of changing the log level is from the Config.toml file that needs to restart the application. We need a new API to change the level programmatically.

### PR DESCRIPTION
**Pull request by daneshk**
## Purpose

Fixes: https://github.com/ballerina-platform/ballerina-library/issues/6213

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec
- [x] Checked native-image compatibility

## Dynamic Log Level Configuration Support

This PR introduces runtime, programmatic control for logger configuration and a registry to manage loggers without restarting the application or editing configuration files.

What changed
- Runtime APIs:
  - Logger.getLevel() to query the effective level.
  - Logger.setLevel(Level) to change levels at runtime (errors on unsupported child-logger operations).
- Logger registry:
  - New LoggerRegistry singleton with getIds() and getById(id) to discover and retrieve loggers.
  - Support for explicit Config.id and auto-generated module:function-based IDs; module-prefixed IDs for module-level loggers.
  - Child loggers inherit parent levels and are not registered independently.
- Native integration:
  - New Java singleton LogConfigManager for logger ID generation and module-level level storage.
  - New native bindings: generateLoggerIdNative, getModuleLevelNative, setModuleLevelNative to coordinate Ballerina ⇄ Java runtime behavior.
- Logging internals:
  - Centralized effective-level evaluation supporting module overrides.
  - New ChildLogger type and refactored printLog flow for context merging, formatting, rotation, and multi-destination writes.
  - Thread-safe registry operations and level updates; module-level changes synchronized with native side.
- Tests, docs and packaging:
  - Extensive tests covering registry behavior, get/set level, inheritance, ID generation, and edge cases.
  - Integration sample and spec/docs updated to document new APIs and Config.id/enableSensitiveDataMasking fields.
  - Version bumps to 2.17.0 and related dependency updates; SpotBugs exclusion added for the singleton.

Performance notes
- Benchmarks show no meaningful regression on the common suppressed-call path (~+0.4%).
- Module-override lookups are faster (≈−7% on suppressed paths, ≈−4% on live paths) due to concurrent registry lookups.
- Registry operations are lightweight (getById ≈40 ns, getIds ≈216 ns); child loggers incur additional cost when merging key-values.

Outcomes
- Enables dynamic adjustment of global and module-specific log levels at runtime via public APIs and native hooks.
- Preserves common-call performance while improving module-override lookup throughput.
- Provides a test-covered, observable API surface and integration points for runtime components to manage logger configuration.

### CODE REVIEW

File: integration-tests/tests/resources/samples/logger/logger-registry/main.bal [hunk 1]
```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2025 WSO2 LLC. (https://www.wso2.com).
```
Comment (daneshk) [inline integration-tests/tests/resources/samples/logger/logger-registry/main.bal]:
```suggestion
// Copyright (c) 2026 WSO2 LLC. (https://www.wso2.com).
```

File: ballerina/init.bal [hunk 2]
```diff
@@ -21,6 +21,23 @@ function init() returns error? {
...
+        loggerRegistry["root"] = rootLogger;
+    }
+
+    // Register each configured module as a Logger in the Ballerina-side registry
+    // Module loggers use the module name as their ID
+    foreach Module mod in modules {
+        ConfigInternal moduleConfig = {
+            level: mod.level
+        };
+        RootLogger moduleLogger = new RootLogger(moduleConfig, mod.name);
```
Comment (TharmiganK) [inline ballerina/init.bal]: How is this going to work? We create new loggers for each configured module but changing the level of that logger will not change the actual logger level right? They will continue to use the root loggers?

File: ballerina/logger.bal [hunk 3]
```diff
@@ -53,4 +53,20 @@ public type Logger isolated object {
...
+   # For root and custom loggers, returns the explicitly set level.
+   # For child loggers (created via `withContext`), returns the inherited level from the parent logger.
+   #
+   # + return - The effective log level
+   public isolated function getLevel() returns Level;
+
+   # Sets the log level of this logger at runtime.
+   # This is supported on root loggers, module loggers, and loggers created via `fromConfig`.
+   # Child loggers (created via `withContext`) do not support this operation and will return an error.
+   # To change a child logger's effective level, set the level on its parent logger instead.
```
Comment (TharmiganK) [inline ballerina/logger.bal:67]: If there is such requirement, how can we get the parent logger of a child logger?

File: ballerina/natives.bal [hunk 4]
```diff
@@ -468,14 +468,17 @@ isolated function replaceString(handle receiver, handle target, handle replaceme
     paramTypes: ["java.lang.CharSequence", "java.lang.CharSequence"]
 } external;
 
-isolated function isLogLevelEnabled(string loggerLogLevel, string logLevel, string moduleName) returns boolean {
```
Comment (TharmiganK) [inline ballerina/natives.bal:471]: Why did we remove this? Don't we handle log level per module?

File: ballerina/root_logger.bal [hunk 5]
```diff
@@ -19,6 +19,9 @@ import ballerina/observe;
 
 # Configuration for the Ballerina logger
 public type Config record {|
+    # Optional unique identifier for this logger.
+    # If provided, the logger will be visible in the ICP dashboard, and its log level can be modified at runtime.
```
Comment (TharmiganK) [inline ballerina/root_logger.bal]: Since this is designed as a common feature, we can avoid mentioning ICP here right?

File: ballerina/root_logger.bal [hunk 6]
```diff
@@ -43,6 +46,39 @@ final string ICP_RUNTIME_ID_KEY = "icp.runtimeId";
...
+        lock {
+            return loggerRegistry.keys().clone();
+        }
+    }
+
+    # Returns a logger by its registered ID.
+    #
+    # + id - The logger ID to look up
+    # + return - The Logger instance if found, nil otherwise
+    public isolated function getById(string id) returns Logger? {
```
Comment (TharmiganK) [inline ballerina/root_logger.bal:78]: Should we design the API like this or should we introduce two APIs:
1. `hasId` - Returns a boolean indicating whether a logger with the specified Id exists or not
2. `getById` - Returns the Logger and panics if there is an error

We tends to use both approaches, just want to confirm whether there is a common practice for Ballerina

File: ballerina/root_logger.bal [hunk 7]
```diff
@@ -43,6 +46,39 @@ final string ICP_RUNTIME_ID_KEY = "icp.runtimeId";
 
 final RootLogger rootLogger;
 
+// Ballerina-side logger registry: loggerId -> Logger
+isolated map<Logger> loggerRegistry = {};
+
+# Provides access to the logger registry for discovering and managing registered loggers.
+public isolated class LoggerRegistry {
```
Comment (TharmiganK) [inline ballerina/root_logger.bal]: Ideally, we should not expose this class right? But I understand the requirement of exposing this APIs. Since we cannot have the init function as private and class as public, how about defining a type for this and use a private object to implement the internal logger registry? With that approach we just expose an interface of the registry and will not allow users to create instances of LoggerRegistry

In that case, we can move the loggerRegistry map to that object and expose APIs to register loggers

File: ballerina/root_logger.bal [hunk 8]
```diff
@@ -60,31 +96,60 @@ public isolated function fromConfig(*Config config) returns Logger|Error {
...
+        // Check for duplicate ID in Ballerina-side registry
+        lock {
+            if loggerRegistry.hasKey(fullId) {
+                return error Error("Logger with ID '" + fullId + "' already exists");
+            }
+        }
+        loggerId = fullId;
+    } else {
+        // No user ID - auto-generate a readable ID
+        loggerId = generateLoggerIdNative(4);
```
Comment (TharmiganK) [inline ballerina/root_logger.bal]: I believe `4` points to the function which has the logger fromConfig call? Better add some comment here for clarity

Also if we define a logger outside a function body i.e. module level, then what will be the prefix we add?

Comment (TharmiganK) [inline ballerina/root_logger.bal]: When auto generating, don't we add the module prefix to the id?

Comment (daneshk) [reply ballerina/root_logger.bal]: For module-level loggers, it will go in the module init function.

Comment (daneshk) [reply ballerina/root_logger.bal]: It is already added when generating the ID.

File: ballerina/logger.bal (hunk 3 above)
Comment (daneshk) [reply ballerina/logger.bal:67]: Inside the child logger class, we keep the parent logger reference.

File: ballerina/init.bal (hunk 2 above)
Comment (daneshk) [reply ballerina/init.bal]: No, moduleLogger is similar to the custom logger, which can only change the log level. We can change the log level of the module logger, and it will change the actual logger's level. The log level is not inherited from the root logger.

**Comment by daneshk** [comment]
# Performance Comparison: `ballerina/log` 2.16.1 vs 2.17.0

## Overview

This document compares the performance of the released `ballerina/log:2.16.1` against the upcoming
`2.17.0` which introduces runtime log level modification via a logger registry.

**Benchmark setup:**
- 2,000,000 iterations × 5 runs per scenario (avg ns/call reported)
- 200,000 iteration warmup before each scenario
- Live log output redirected to file to exclude I/O variance from suppressed-call benchmarks
- Machine: macOS Darwin 25.2.0

---

## Scenario 1 — No Module Override, Suppressed Call

> Root level = `INFO`. `printDebug` is suppressed. Exercises the level-check hot path with no
> module-level registry hit.

| Metric | 2.16.1 | 2.17.0 | Diff |
|--------|--------|--------|------|
| `log:printDebug` (suppressed) | 1692 ns | 1699 ns | +7 ns (+0.4%) |

---

## Scenario 2 — No Module Override, Live Call

> Root level = `INFO`. `printInfo` is live. Exercises the full print path including formatting
> and file I/O.

| Metric | 2.16.1 | 2.17.0 | Diff |
|--------|--------|--------|------|
| `log:printInfo` (live) | 28,061 ns | 29,022 ns | +961 ns (+3.4%) |

---

## Scenario 3 — Module-Level Override Present, Suppressed Calls

> `Config.toml` sets the benchmark module level to `WARN`. `printInfo` and `printDebug` are
> suppressed via the module-level override. Exercises the registry lookup path (hit).

| Metric | 2.16.1 | 2.17.0 | Diff |
|--------|--------|--------|------|
| `log:printInfo` (INFO \< WARN, suppressed) | 2,184 ns | 2,041 ns | **−143 ns (−6.5%)** ✅ |
| `log:printDebug` (DEBUG \< WARN, suppressed) | 2,182 ns | 2,011 ns | **−171 ns (−7.8%)** ✅ |

> **Note:** 2.17.0 is faster here because the Java `ConcurrentHashMap.get()` outperforms the
> previous native module-level check used in 2.16.1.

---

## Scenario 4 — Module-Level Override Present, Live Call

> `Config.toml` sets the benchmark module level to `WARN`. `printWarn` is live via the
> module-level override.

| Metric | 2.16.1 | 2.17.0 | Diff |
|--------|--------|--------|------|
| `log:printWarn` (WARN \>= WARN, live) | 29,550 ns | 28,397 ns | **−1,152 ns (−3.9%)** ✅ |

---

## Scenario 5 — 2.17.0 New API Scenarios

> These scenarios exercise APIs introduced in 2.17.0: `fromConfig`, `setLevel`, `getLevel`,
> `root()`, `getLoggerRegistry()`, and `withContext` child loggers.

| Scenario | ns/call |
|----------|---------|
| `fromConfig` logger — suppressed DEBUG (level=INFO) | 1,747 ns |
| `fromConfig` logger — live DEBUG (level=DEBUG) | 19,502 ns |
| `root()` API — suppressed DEBUG | 1,718 ns |
| `root().setLevel(DEBUG)` then live `log:printDebug` | 18,230 ns |
| `fromConfig` + `setLevel(DEBUG)` — suppressed (before set) | 1,622 ns |
| `fromConfig` + `setLevel(DEBUG)` — live (after set) | 18,414 ns |
| `getLoggerRegistry().getById("root")` | **40 ns** |
| `getLoggerRegistry().getIds()` | **216 ns** |
| `withContext` child logger — suppressed DEBUG | 3,903 ns |
| `withContext` child logger — live DEBUG | 24,989 ns |

---

## Summary

| Scenario | 2.16.1 | 2.17.0 | Verdict |
|----------|--------|--------|---------|
| Suppressed call, no module override | 1,692 ns | 1,699 ns | ✅ No regression (+0.4%) |
| Live call, no module override | 28,061 ns | 29,022 ns | ✅ Acceptable (+3.4%) |
| Suppressed call, module override hit | 2,183 ns | 2,026 ns | ✅ Faster (−7.2%) |
| Live call, module override hit | 29,550 ns | 28,397 ns | ✅ Faster (−3.9%) |
| Registry `getById` | — | 40 ns | ✅ Very fast |
| Registry `getIds` | — | 216 ns | ✅ Fast |
| `fromConfig` suppressed | — | 1,747 ns | ✅ Comparable to root |
| `fromConfig` live | — | 19,502 ns | ✅ Comparable to root |
| `withContext` child suppressed | — | 3,903 ns | ⚠️ 2× root (key-value merge overhead) |

### Key Findings

1. **No meaningful regression for existing use cases.** The suppressed-call hot path (the most
   common case in production) shows only a +0.4% difference with no module overrides configured.

2. **Module-level overrides are faster in 2.17.0.** The Java `ConcurrentHashMap.get()` used for
   the registry lookup outperforms the previous native implementation by ~7% on suppressed calls
   and ~4% on live calls.

3. **Registry lookups are cheap.** `getById` costs only 40 ns and `getIds` 216 ns, making
   runtime logger discovery negligible.

4. **`fromConfig` loggers perform on par with the root logger** on both the suppressed and live
   paths.

5. **`withContext` child loggers carry ~2× overhead** on suppressed calls (3,903 ns vs ~1,700 ns)
   due to the key-value merge step. This is expected and consistent with the previous behaviour.

6. **The dominant cost on the suppressed path is the JVM stack walk** (`getInvokedModuleName`),
   which accounts for ~1,500 ns regardless of version. Lock and registry overhead are negligible
   in comparison.

File: ballerina/root_logger.bal (hunk 6 above)
Comment (daneshk) [reply ballerina/root_logger.bal:78]: I prefer this approach rather than panic. So the application won't break if the user enters the wrong ID

File: ballerina/root_logger.bal (hunk 8 above)
Comment (daneshk) [reply ballerina/root_logger.bal]: Add comment for the stack offset 4

File: ballerina/natives.bal (hunk 4 above)
Comment (daneshk) [reply ballerina/natives.bal:471]: It is there in the print function https://github.com/daneshk/module-ballerina-log/blob/ca4cc160f86b660e1ef0e149d3c814be191783e5/ballerina/root_logger.bal#L212-L221

File: native/src/main/java/io/ballerina/stdlib/log/LogConfigManager.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            // The version segment (e.g., "0") and file name should be stripped.
+            String[] parts = className.split("\\.");
+            if (parts.length >= 3) {
+                // parts[0] = org, parts[1] = module, parts[2] = version, parts[3..] = file/class
+                modulePart = parts[0] + "/" + parts[1];
+            } else if (parts.length == 2) {
+                modulePart = parts[0] + "/" + parts[1];
+            } else {
+                modulePart = className;
+            }
```
[the remark this answers is not in the document]
Comment (daneshk) [reply native/src/main/java/io/ballerina/stdlib/log/LogConfigManager.java:101]: Fixed in [ef8156f](https://github.com/ballerina-platform/module-ballerina-log/commit/ef8156fd3a86e41429715aeeb55e29f663f7833f)

File: native/src/main/java/io/ballerina/stdlib/log/LogConfigManager.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                        versionIdx = i;
+                        break;
+                    }
+                }
+                // Module name is everything from parts[1] up to (not including) the version segment
+                String rawModule = String.join(".", Arrays.copyOfRange(parts, 1, versionIdx));
+                modulePart = parts[0] + "/" + IdentifierUtils.decodeIdentifier(rawModule);
+            } else {
+                modulePart = className;
+            }
```
[the remark this answers is not in the document]
Comment (daneshk) [reply native/src/main/java/io/ballerina/stdlib/log/LogConfigManager.java:101]: Fixed in 3898ca6

File: ballerina/root_logger.bal [hunk 11]
```diff
@@ -112,88 +177,153 @@ isolated class RootLogger {
...
-                select element.toString();
+    }
+
+    isolated function print(string logLevel, string moduleName, string|PrintableRawTemplate msg, error? err = (), error:StackFrame[]? stackTrace = (), *KeyValues keyValues) {
+        if !isLevelEnabled(self.getLevel(), logLevel) {
+            return;
         }
+        printLog(logLevel, moduleName, msg, self.format, self.destinations, self.keyValues,
+                self.enableSensitiveDataMasking, err, stackTrace, keyValues);
+    }
```
[the remark this answers is not in the document]
Comment (daneshk) [reply ballerina/root_logger.bal:222]: It is there in the print function https://github.com/daneshk/module-ballerina-log/blob/ca4cc160f86b660e1ef0e149d3c814be191783e5/ballerina/root_logger.bal#L212-L221

File: native/src/main/java/io/ballerina/stdlib/log/LogConfigManager.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            String[] parts = className.split("\\.");
+            if (parts.length >= 3) {
+                // parts[0] = org, parts[1] = module, parts[2] = version, parts[3..] = file/class
+                modulePart = parts[0] + "/" + parts[1];
+            } else if (parts.length == 2) {
+                modulePart = parts[0] + "/" + parts[1];
+            } else {
+                modulePart = className;
+            }
+            functionPart = methodName;
```
[the remark this answers is not in the document]
Comment (daneshk) [reply native/src/main/java/io/ballerina/stdlib/log/LogConfigManager.java:102]: This is fixed in [ef8156f](https://github.com/ballerina-platform/module-ballerina-log/pull/1447/commits/ef8156fd3a86e41429715aeeb55e29f663f7833f)

File: ballerina/natives.bal (hunk 4 above)
Comment (daneshk) [reply ballerina/natives.bal:471]: Runtime level change

  When someone does check moduleLogger.setLevel(DEBUG) (root_logger.bal:201):
```  
public isolated function setLevel(Level level) returns error? {
      lock {
          self.currentLevel = level;        // ← Ballerina lock, updates RootLogger field
      }
      if self.isModuleLogger {
          setModuleLevelNative(self.loggerId ?: "", level);  // ← updates Java ConcurrentHashMap
      }
  }
```

  Both the Ballerina object field and the Java map are updated atomically (each in their own mechanism). The ConcurrentHashMap is the source of truth for the hot print path.

File: ballerina/root_logger.bal [hunk 13]
```diff
@@ -112,86 +195,188 @@ isolated class RootLogger {
...
+        }
+        return new ChildLogger(self, newKeyValues.cloneReadOnly());
+    }
+
+    public isolated function getLevel() returns Level {
+        return self.parent.getLevel();
+    }
+
+    public isolated function setLevel(Level level) returns error? {
+        return error("Unsupported operation: cannot set log level on a child logger. " +
```
Comment (TharmiganK) [inline ballerina/root_logger.bal]:
```suggestion
        return error Error("Unsupported operation: cannot set log level on a child logger. " +
```

**Review by ayeshLK — APPROVED** [review]

**Review by TharmiganK — APPROVED** [review]
